server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -1481,28 +1481,33 @@ public static DataStreamLifecycle resolveLifecycle(
      * [
      *   {
      *     "lifecycle": {
+     *       "enabled": true,
      *       "data_retention" : "10d"
      *     }
      *   },
      *   {
      *     "lifecycle": {
+     *       "enabled": true,
      *       "data_retention" : "20d"
      *     }
      *   }
      * ]
-     * The result will be { "lifecycle": { "data_retention" : "20d"}} because the second data retention overrides the first.
-     * However, if we have the following two lifecycles:
+     * The result will be { "lifecycle": { "enabled": true, "data_retention" : "20d"}} because the second data retention overrides the
+     * first. However, if we have the following two lifecycles:
      * [
      *   {
      *     "lifecycle": {
+     *       "enabled": false,
      *       "data_retention" : "10d"
      *     }
      *   },
      *   {
-     *   "lifecycle": { }
+     *   "lifecycle": {
+     *      "enabled": true
+     *   }
      *   }
      * ]
-     * The result will be { "lifecycle": { "data_retention" : "10d"} } because the latest lifecycle does not have any
+     * The result will be { "lifecycle": { "enabled": true, "data_retention" : "10d"} } because the latest lifecycle does not have any
```

Comment:
Sorry, I do not fully understand the comment here. A lifecycle that is not enabled is disabled `"lifecycle": {"enabled": false}`, right? So, it should influence the composition because it is the "direct order" of the user to disable the lifecycle.

Reply:
Sorry, maybe my comment was misplaced in the PR. 

> The result will be { "lifecycle": { "enabled": true, "data_retention" : "10d"} }

This is the final result after composition. The composed lifecycles were :
```
  "lifecycle": {
      "enabled": false,
      "data_retention" : "10d"
    }
  },
   "lifecycle": {
   "enabled": true
 }
```
The `data_rentetion` is present in in the final result because it was configured in a lifecycle object that is disabled (the first one : `enabled: false`)
I find this behaviour strange, as a disabled lifecycle should not (IMO) influence the output of the composition.

Reply:
Ah, now I see! I have been doubting about this myself. In other words, when we have `disabled: false` should we not allow the user to set retention and downsampling. This would align with what you said earlier that we do not keep the retention from:
```
"lifecycle": {
      "enabled": false,
      "data_retention" : "10d"
    }
```
But if we do not keep it during composition, I think we should disallow this configuration to begin with and require that all other configuration is not set when `enabled: false`. Do you think this makes more sense?


The only downside, I can think of would be when a user wants to temporarily disable  the lifecycle without losing the existing configuration, for maintenance reasons, for example. Not sure  though if this is a real need.

Reply:
>  In other words, when we have disabled: false should we not allow the user to set retention and downsampling.

I think we should continue to keep the entire lifecycle configuration. 

> I can think of would be when a user wants to temporarily disable the lifecycle without losing the existing configuration, for maintenance reasons, for example. Not sure though if this is a real need.

I think we should continue to support this.

However, in composition, I think `enabled: false` should be semantically interpreted and the _other_ lifecycle fields (except `enabled` ) should not participate in the composition anymore.

Reply:
I have to admit I find this differentiation a bit confusing, why would we treat it differently in composition than when it's set on a data stream level?

Reply:
> why would we treat it differently in composition than when it's set on a data stream level?

I don't think it'd be treated differently. In both cases the lifecycle would have no effect. 
If it's configured on a data stream and subsequently disabled, it'll have no effect (the data stream would be un-managed). If it was later re-enabled, the data stream would be managed and the configuration would start taking effect. Note we can't just change the `enabled` field but when we modify a lifecycle we have to provide the entire lifecycle configuration - so when re-enabling a lifecycle it'll have to be:
``` 
   "lifecycle": {
      "enabled": true,
      "data_retention" : "10d"
    }
```
Keeping the `data_retention` in the configuration even when the lifecycle is disabled could help the UI build an experience where it _seems_ like you're just changing the `enabled` field (because you know what was there before).

In index template composition, ignoring ` "data_retention" : "10d"` when the `lifecycle` is NOT `enabled` is being consistent IMO (i.e. a lifecycle that's not enabled doesn't influence the outcome, except by saying that it's not enabled for opting-out purposes)

Reply:
I think we should treat each individual configuration option separately to avoid the action at a distance, meaning Mary's original composition method. The reason is that it's more apparent to the user what the behavior is, rather than a special case for composition. That means each key (`data_retention`, `enabled`, `downsampling`) is treated just like a regular key in a map, being overwritten on merging.

Reply:
After an offline discussion with @andreidan we agreed that we will keep the composition as is, we agree with the way @dakrone summarised it:

> ... it's more apparent to the user what the behavior is, rather than a special case for composition. That means each key (data_retention, enabled, downsampling) is treated just like a regular key in a map, being overwritten on merging.